### PR TITLE
Spatial index by product

### PR DESCRIPTION
**Pull request by SpacemanPaul**
### Reason for this pull request

Add product and dataset resource spatial and temporal extent methods.

This probably should have been included in #1538 - but we shook out a better API design, so all good.

### Proposed changes

- Add optional "product" argument to `spatial_extent()` method, for consistency with `temporal_extent()`, as per [EP13](https://github.com/opendatacube/datacube-core/wiki/ODC-EP-013-Index-Driver-API-cleanup).

 - [x] Tests added / passed
 - [x] Fully documented, including `docs/about/whats_new.rst` for all changes

### CODE REVIEW

**Comment by omad** [comment]
Thanks Paul,

This looks like a useful feature, but I'm not so keen on the API. 

I think it is better to not overload functions with optional arguments that completely change what the function does. I'd prefer separate methods for 'extents of a list of datasets' to 'extents of a product'.

In doing so, I think it would even make sense to move the 'product extents', methods to the `ProductResource` instead of the `DatasetResource`.

Happy to discuss. :)

**Comment by SpacemanPaul** [comment]
> Thanks Paul,
> 
> This looks like a useful feature, but I'm not so keen on the API.
> 
> I think it is better to not overload functions with optional arguments that completely change what the function does. I'd prefer separate methods for 'extents of a list of datasets' to 'extents of a product'.
> 
> In doing so, I think it would even make sense to move the 'product extents', methods to the `ProductResource` instead of the `DatasetResource`.
> 
> Happy to discuss. :)

That makes total sense to me. It complicates the deprecation warnings in the interim 1.9 stage of the migration a bit, but nothing that need get in our way.  I'll update EP13 and this PR this afternoon.

(The argument for leaving in the `DatasetResource` would be that all current drivers implement by reading all datasets for a product, so although the API specifies a product, the actual database operation are happening on the dataset table.  But in future, index drivers may choose to maintain a product-level extent index, in which scenario this distinction disappears, which in turns tells me this is a bad argument.)

File: datacube/index/abstract.py [hunk 1]
```diff
@@ -779,6 +779,34 @@ def get_all_docs(self) -> Iterable[Mapping[str, Any]]:
         for prod in self.get_all():
             yield prod.definition
 
+    @abstractmethod
+    def spatial_extent(self, product: str | Product, crs: CRS = CRS("EPSG:4326")) -> Optional[Geometry]:
+        """
+        Return the combined spatial extent of the nominated datasets (or all datasets for the nominated product)
```
Comment (Ariana-B) [inline datacube/index/abstract.py]: Docstring needs updating

File: datacube/index/abstract.py [hunk 2]
```diff
@@ -779,6 +779,34 @@ def get_all_docs(self) -> Iterable[Mapping[str, Any]]:
...
+
+        :param product: A Product or product name. (or None)
+        :param crs: A CRS (defaults to EPSG:4326)
+        :return: The combined spatial extents of the product.
+        """
+
+    @abstractmethod
+    def temporal_extent(self, product: str | Product) -> tuple[datetime.datetime, datetime.datetime]:
+        """
+        Returns the minimum and maximum acquisition time of a product or an iterable of dataset ids.
```
Comment (Ariana-B) [inline datacube/index/abstract.py]: Docstring is incorrect

File: datacube/index/abstract.py [hunk 3]
```diff
@@ -1696,24 +1724,18 @@ def search_eager(self, **query: QueryField) -> List[Dataset]:
         return list(self.search(**query))  # type: ignore[arg-type]   # mypy isn't being very smart here :(
 
     @abstractmethod
-    def temporal_extent(self,
-                        product: str | Product | None,
-                        ids: Iterable[DSID] | None
-                        ) -> tuple[datetime.datetime, datetime.datetime]:
+    def temporal_extent(self, ids: Iterable[DSID]) -> tuple[datetime.datetime, datetime.datetime]:
         """
         Returns the minimum and maximum acquisition time of a product or an iterable of dataset ids.
```
Comment (Ariana-B) [inline datacube/index/abstract.py]: Docstring needs to be updated

File: datacube/index/abstract.py [hunk 4]
```diff
@@ -1762,15 +1784,16 @@ def search_returning_datasets_light(self,
...
-        Return the combined spatial extent of the nominated datasets.
+        Return the combined spatial extent of the nominated datasets
 
         Uses spatial index.
+
         Returns None if no index for the CRS, or if no identified datasets are indexed in the relevant spatial index.
         Result will not include extents of datasets that cannot be validly projected into the CRS.
 
-        :param ids: An iterable of dataset IDs
+        :param ids: An iterable of dataset IDs (or None)
```
Comment (Ariana-B) [inline datacube/index/abstract.py]: `ids` param type doesn't look to be optional

File: datacube/index/memory/_products.py [hunk 5]
```diff
@@ -19,8 +20,9 @@
 
 
 class ProductResource(AbstractProductResource):
-    def __init__(self, metadata_type_resource):
-        self.metadata_type_resource: MetadataTypeResource = metadata_type_resource
+    def __init__(self, index):
+        self._index = index
+        self.metadata_type_resource: MetadataTypeResource = index.metadata_types
```
Comment (Ariana-B) [inline datacube/index/memory/_products.py]: why have this when it seems we're now using `self._index.metadata_types` instead? If there is a difference between the two and their use cases, it should be clarified

File: datacube/index/null/_datasets.py [hunk 6]
```diff
@@ -106,23 +106,12 @@ def count_product_through_time(self, period, **query):
...
-            raise KeyError(str(product))
+    def temporal_extent(self, ids: Iterable[DSID]) -> tuple[datetime.datetime, datetime.datetime]:
+        raise KeyError(str(ids))
 
     # pylint: disable=redefined-outer-name
     def search_returning_datasets_light(self, field_names: tuple, custom_offsets=None, limit=None, **query):
         return []
 
-    def spatial_extent(self, ids, crs=None):
+    def spatial_extent(self, ids=None, product=None, crs=None):
```
Comment (Ariana-B) [inline datacube/index/null/_datasets.py]: I thought product param was removed?

File: datacube/index/postgis/_datasets.py [hunk 7]
```diff
@@ -910,7 +887,14 @@ def get_custom_query_expressions(self, custom_query, custom_offsets):
 
         return custom_exprs
 
-    def spatial_extent(self, ids: Iterable[DSID], crs: CRS = CRS("EPSG:4326")) -> Optional[Geometry]:
+    def temporal_extent(self, ids: Iterable[DSID]) -> tuple[datetime.datetime, datetime.datetime]:
+        """
+        Returns the minimum and maximum acquisition time of the product.
```
Comment (Ariana-B) [inline datacube/index/postgis/_datasets.py]: of the ids, not product

File: datacube/index/postgres/_datasets.py [hunk 8]
```diff
@@ -738,6 +736,9 @@ def search_summaries(self, **query):
             for columns in results:
                 yield columns._asdict()
 
+    def spatial_extent(self, ids, crs=None):
+        return None
+
     def temporal_extent(
             self,
             product: str | Product | None = None,
```
Comment (Ariana-B) [inline datacube/index/postgres/_datasets.py]: shouldn't this param be removed?

**Review by Ariana-B — CHANGES_REQUESTED** [review]
Looks reasonable overall, but needs some cleanup

File: datacube/index/abstract.py (hunk 4 above)
Comment (SpacemanPaul) [reply datacube/index/abstract.py]: Thanks for picking up all these.

File: datacube/index/memory/_products.py (hunk 5 above)
Comment (SpacemanPaul) [reply datacube/index/memory/_products.py]: Yeah, just me being lazy.  I'll fix it.

**Review by Ariana-B — APPROVED** [review]
lgtm

**Pull request merged**